wait_enqueue

koraa opened the pull request:
Hi,

I would like to propose adding a family of wait_enqueue methods and close method to the Blocking concurrent queue.

Blocking writes are useful to throttle producers; e.g. if producers are much faster than consumers, producers will simply made to wait until there are capacities on the consumers.

cameron314 commented on blockingconcurrentqueue.h:
```diff
@@ -362,6 +373,19 @@ namespace details
+      ssize_t tryWaitManyExclusive(ssize_t max) {
+				assert(max >= 0);
+				ssize_t oldCount = m_count.load(std::memory_order_relaxed);
+        while (true) {
```
This loop can be simplified to:

    while (oldCount >= max) {
        if (m_count.compare_exchange_weak(oldCount, oldCount - max, std::memory_order_acquire, std::memory_order_relaxed))
            return max;
    }
    return 0;

cameron314 commented on blockingconcurrentqueue.h:
```diff
@@ -452,19 +476,23 @@ class BlockingConcurrentQueue
 	BlockingConcurrentQueue(size_t minCapacity, size_t maxExplicitProducers, size_t maxImplicitProducers)
-		: inner(minCapacity, maxExplicitProducers, maxImplicitProducers), sema(create<LightweightSemaphore>(), &BlockingConcurrentQueue::template destroy<LightweightSemaphore>)
+		: inner(minCapacity, maxExplicitProducers, maxImplicitProducers),
+      sema_dequeue(create<LightweightSemaphore>(), &BlockingConcurrentQueue::template destroy<LightweightSemaphore>),
```
Nitpick: Indentation

cameron314 commented on blockingconcurrentqueue.h:
```diff
@@ -506,7 +536,9 @@ class BlockingConcurrentQueue
-		sema.swap(other.sema);
+		sema_dequeue.swap(other.sema_dequeue);
+		sema_enqueue.swap(other.sema_enqueue);
+    std::swap(closed_, other.closed_);
```
Indentation

cameron314 commented on blockingconcurrentqueue.h:
```diff
@@ -604,8 +636,11 @@ class BlockingConcurrentQueue
-		if (inner.try_enqueue(item)) {
-			sema->signal();
+    if (sema_enqueue->tryWait()) {
+      while (!inner.try_enqueue(item)) {
```
Nitpick: Unnecessary braces

cameron314 commented on blockingconcurrentqueue.h:
```diff
@@ -452,19 +476,23 @@ class BlockingConcurrentQueue
-		: inner(capacity), sema(create<LightweightSemaphore>(), &BlockingConcurrentQueue::template destroy<LightweightSemaphore>)
+		: inner(capacity),
+      sema_dequeue(create<LightweightSemaphore>(), &BlockingConcurrentQueue::template destroy<LightweightSemaphore>),
+      sema_enqueue(create<LightweightSemaphore>(), &BlockingConcurrentQueue::template destroy<LightweightSemaphore>)
```
`sema_enqueue` starts off with a size `0` capacity; it should start with the number of free slots instead. However, since the actual number of free slots varies depending on `BLOCK_SIZE` and whether an implicit or explicit producer is being queued to, it's actually impossible to know whether an `enqueue` will succeed or not unless the maximum number of concurrent implicit & explicit producers are known (to put a bound on the available number of slots at any given time). Even then, there's usually several wasted blocks. I'm not sure how this would work...

cameron314 commented on blockingconcurrentqueue.h:
```diff
@@ -519,7 +551,7 @@ class BlockingConcurrentQueue
 	inline bool enqueue(T const& item)
```
Shouldn't this reduce the count of the semaphore?

koraa replied:
Well, my thinking here was that enqueue() must never fail, so in effect this would probably increase the amount of enqueue-slots available in the underlying buffer.
By not decreasing the count here, it will implicitly increase the enqueue semaphore later when the slot becomes free again from dequeueing, accounting for the extra slot that became available in the queue.

Although this line of thinking is a bit weird.
I think I'll probably instead just decrease the semaphore; I'll have to unconditionally decrease beyond zero though to not destroy the never-fails property of enqueue()

cameron314 replied:
Ah, I think I understand, at least the first part.
Unfortunately, `enqueue` can still fail for various reasons, e.g. if there's not enough memory. But even if it succeeds, it does not necessarily create a new slot -- it tries to use an existing block, if any, then requests a new block from the queue if that fails, which may give it a recycled or preallocated block. A new block is allocated only if that also fails. (And a block has more than one slot.)

So if there's one call the `enqueue` and another to `try_enqueue`, the `try_enqueue` might fail if the `enqueue` took the last slot even though if called in the opposite order they would both succeed since `try_enqueue` would end up allocating a new block.

cameron314 commented on blockingconcurrentqueue.h:
```diff
@@ -674,25 +721,309 @@ class BlockingConcurrentQueue
+  // If enough room could be acquired, the element is enqueued (by moving it) and true is returned.
+  // If the timeout expired before the element could be enqueued, false is returned.
+  // Never allocates. Thread-safe.
+	inline bool wait_enqueue_timed(T &&item, std::int64_t timeout_usecs)
```
Indentation

cameron314 commented on blockingconcurrentqueue.h:
```diff
@@ -674,25 +721,309 @@ class BlockingConcurrentQueue
+	// Does not allocate memory (except for one-time implicit producer).
+  // Thread-safe.
+	template<typename It>
+  inline void wait_enqueue_bulk(It itemFirst, size_t max)
```
`max` should be named `count` instead.

cameron314 commented on blockingconcurrentqueue.h:
```diff
@@ -674,25 +721,309 @@ class BlockingConcurrentQueue
+    size_t count = 0;
+    while (count < max) {
+		  size_t gained = (size_t)sema_enqueue->waitMany((LightweightSemaphore::ssize_t)(ssize_t)(max - count));
+      while (!inner.try_enqueue_bulk(std::forward<It&>(itemFirst), gained)) {
```
Can you verify that this succeeds on the first implicit producer allocation?

cameron314 replied:
Does `itemFirst` increment itself automatically?

cameron314 commented on blockingconcurrentqueue.h:
```diff
@@ -674,25 +721,309 @@ class BlockingConcurrentQueue
+  {
+    size_t count = 0;
+    while (count < max) {
+		  size_t gained = (size_t)sema_enqueue->waitMany((LightweightSemaphore::ssize_t)(ssize_t)(max - count));
```
Not sure this is a good idea. It could lead to deadlock where two producers each acquire half the number of slots they need and are waiting on each other. Would be better to avoid a loop and use a single `waitManyExact` that waits for the entire count to become free.

koraa replied:
Hmm, I did think of that; the reason I ultimately decided to do it this because otherwise I could either only gurantee that *some* items are enqueued or I could never enqueue all of the elements passed. I'll go for the weaker variant, not guranteeing that all elements will be enqueued then…

cameron314 commented on blockingconcurrentqueue.h:
```diff
@@ -674,25 +721,309 @@ class BlockingConcurrentQueue
 			while (!inner.try_dequeue(item)) {
 				continue;
 			}
+			sema_enqueue->signal();
```
Because elements are stored in blocks, dequeueing an item doesn't free up a slot for enqueueing right away. Not sure how to reconcile this.

koraa replied:
Well, the enqueue ops will spin so I think it would be work, but it's not very efficient :/

cameron314 replied:
They could end up spinning indefinitely if only one item is ever dequeued :P

cameron314 commented on blockingconcurrentqueue.h:
```diff
@@ -897,29 +1283,69 @@ class BlockingConcurrentQueue
+  }
+
+  // Whether close() has been called on this queue.
+  // When the queue is closed, not enqueue calls must be made.
```
Typo. Also, the queue should enforce this (even if it's only in !NDEBUG for performance).

koraa replied:
I though for quite a while about how to implement this, but TBH I could not come up with a working scheme for this, either using an extra atomic counter for the number of current enqueue operations or just letting the enqueue ops fail if the queue was closed, but I am not sure how to do that reliably either.
In the end I always found some cases where finish() might be true for a short time and that's not acceptable.

koraa replied:
If wait_enqueue can fail though that could also lead to some weird cases where bulk waiting enqueues insert some but not all elements…

cameron314 commented on blockingconcurrentqueue.h:
```diff
@@ -897,29 +1283,69 @@ class BlockingConcurrentQueue
+	// Thread-safe.
+  inline void close()
+  {
+	  bool old_closed = closed_.exchange(true, std::memory_order_relaxed, std::memory_order_relaxed);
```
Should use `std::atomic_flag` instead of CAS loop. Relaxed memory ordering seems odd too, are you sure that is correct?

cameron314 commented on blockingconcurrentqueue.h:
```diff
@@ -897,29 +1283,69 @@ class BlockingConcurrentQueue
+  // This signals that no more data will be enqueued to dequeue operations.
+  // Further dequeue operations will behave normally until the queue is
+  // completely drained (finished() == true). After that, any dequeue calls
+  // will immediately fail and return.
```
How? The semaphore says an extra item is available; the `wait_dequeue` methods will do the wrong thing, and the plain `dequeue` methods don't check the flag...

koraa replied:
My idea here was to use a sort of virtual sentinel element.

```
while (!inner.try_dequeue(item)) {
  // This is supposed to check for the sentinel;
  // there is an element 'in' the semaphore, all
  // enqueue ops *must* have stabilized (because
  // close() demands so), so  the only case where
  // dequeue could fail, is if the inner queue is
  // truly depleted.
  if (closed()) {
    // Add a new sentinel element, so the same logic
    // will work again if wait_dequeue is called again.
    sema_dequeue->signal();
    return false;
  }
  ...
}
```


And you are right, I completely forgot to update the logic in
the old dequeue functions!

koraa commented:
Wow, thanks for the comments. I'll fix the problems and add some tests tomorrow ^^

cameron314 commented:
Sure :-)
Sorry if my review seems a bit harsh, I really appreciate the effort. It's just that I'm not sure what you're trying to do with the blocking enqueue is possible given the current queue architecture.

koraa commented:
It's ok. I don't claim to have done something like that before so the review does not seem very harsh…
Since I know there is interest in this from your side I can now also take more than a single day to try to get this right!
About the initial value for the enqueue semaphore, I was always thinking of that it would be nice to actually just specify the number of elements that can at most be enqueued independantly of the number of producers.
Like at the start you could just specify a maximum capacity of 200 elements, which would be perfectly fine if the main motivation is to use the queue for flow control…

koraa commented:
(To clarify my badly formulated thoughts: I never did something like this before, so I expected the result to be pretty bad. I am actually surprised you even took the time to thoroughly review it, so thank you!)

cameron314 commented:
I think what most people really want is a good ring-buffer implementation. My queue is overcomplexified in the name of performance and this causes all sorts of little subtle limitations that people keep running into, including on the implementation side. Sigh.

But yes, if someone submits a juicy pull request of course I'll review it ;-)
I'm not sure if this will end up being merged though. We'll see! I've actually only done a cursory review so far. A full review would look carefully into the memory barriers, etc. and audit every public function to make sure none are broken by the new semaphore.

koraa commented:
Hi,

I just addressed some of your comments. I also got rid of the close() part of this PR.

From what I can see, there are still three basic issues with the code in this PR:
1. enqueue() methods should decrease the sema_enqueue unconditionally, but without waiting. (basically this would circumvent the size limit for a moment.
2. What should sema_enqueue be initialized too?
3. And finally the problem that a slot might take some time to be freed after dequeue().

1. It might be possible to resolve this by implementing an emulated 'SignedSemaphore'. This could be implemented as a subclass of LightSemaphore in which has an extra std::atomic<size_t> or std::atomic<int>; to decrease the semaphore, one could use try_wait() and if that fails just increase the counter. Future signal() calls would subtract the value in the atomc counter from the number of values to signal.
2. For `BlockingConcurrentQueue(size_t capacity)` this should be pretty easy: just initialize the counter to capacity. For `BlockingConcurrentQueue(size_t minCapacity, size_t maxExplicitProducers, size_t maxImplicitProducers)` could be initialized to `minCapacity`
3. I think it should be possible to just define wait_enqueue in terms of enqueue() instead of try_enqueue() in the beginning and depending on the number of producers (implicit or explicit), but eventually it should stabilize to the number of buckets to actually store the items PLUS some more so we can avoid memory allocations in the future.

I am right now moving those three ideas in my had; using enqueue() instead of try_enqueue() seems like a decent idea because it would basically auto-optimize as far as I can see.
Generally though, the solutions to the other two problems seem like slightly cringeworthy hacks; I will have to think about this for a bit but it seems to me – while I am pretty sure that a queue with a capacity could be implemented on top of ConcurrentQueue – the interface of BlockingConcurrentQueue would not work very well with that.
I'll have to turn that thing in my had for a while though… :)

What do you think?

cameron314 commented:
Hmm, yeah, the more I think about it, the more it seems like a blocking enqueue (while a nice feature) can't really be implemented on top of the existing `ConcurrentQueue` architecture.

Here's some responses to your points. I wrote them as I went (and I thought more as I went too), so my response to point 6 is perhaps the most interesting, so start there :-)

1. I missed this subtlety the first time around; I guess `enqueue` would have to unconditionally decrement the semaphore, but also increment it when a new block is allocated.
2. It could be initialized to the number of slots that are pre-allocated, but since slots are claimed wholesale by block in each sub-queue, this doesn't completely make sense either; some sort of upper-bound system would have to be devised that would end up wasting many slots just in padding (this is basically what I do for the pre-allocation in the constructor that takes three arguments -- the comments there are relevant -- but simpler).
3. This is the killer one, I think. The explicit sub-queues never even recycle their blocks back to the central pool at all, so even if a block is completely emptied, it can't be used by a different producer -- yet the semaphore is shared between all producers. Eek!
4. The lightweight semaphore can already take on a negative value, but it represents the number of tasks waiting on the semaphore. If the semaphore is changed to maintain a steady negative value, what criteria blocks and resumes waiters? If an additional value is added to the mix, it will be very difficult to maintain correctness under all concurrent conditions, since there will likely be occasions where both values need to be modified in a single atomic transaction (which they cannot be).
5. I suppose, but the other inter-related problems still stand.
6. I think fundamentally, the idea of `wait_enqueue` is that it should block until there's space in the queue to add the item without allocation (but note that the producer itself could still need allocation. Hmm). Together with an initial capacity, this bounds the number of items in the queue. Perhaps this could work if the bounding constraint was shared between the central queue and each sub-queue instead of attempting to keep it completely centralized with a single semaphore. Each sub-queue would have a semaphore with the number of blocks it has free, and maintain that count accordingly (this is fairly cheap since it's at the block level). A blocking enqueue would first check the current block (to see if there was room left there), then check the semaphore, and if the claimed blocks were already depleted attempt to get another one from the parent central queue object (which would affect the count). The central parent queue, in turn, would have a semaphore on the number of blocks in its pre-allocated/free block pool. This sounds really tricky to get right, though.

koraa commented:
So, I thought a while about the implementation you proposed in (6) and I think I don't want to try it tbh; it would require merging the code of blockingconcurrentqueue into concurrentqueue or using implementation details of concurentqueue; to me it simply seems to hard implement correctly right now.

"the idea of wait_enqueue is that it should block until there's space in the queue to add the item without allocation" – Yeah, I agree with that too; otherwise the semantics of try_enqueue would not match the enqueue semantics and that would be extremely weird.

So, even though that extending concurrentqueue with blocking enqueue (or close for that matter) is not right now, I still think that having a bounded queue with close signaling would be useful.
I think the biggest use for such a data structure could be derived for making code more simple (through close() and waiting) especially for applications using a processing pipline pattern (groups of worker threads passing data through multiple processing stages) without any realtime/high level of control requirement.
In order to provide an implementation for that use case, it might be worth to implement a ConcurrentBoundedQueue similar to BlockingConcurrentqueue, with less rigorous guarantees in regards to memory allocations. It could drop the  `BlockingConcurrentQueue(size_t, size_t, size_t)` constructor as well as the group of always-succeeds `enqueue` operations.
Both `try_enqueue` as well as `wait_enqueue` could be implemented in terms of `Concurrentqueue::enqueue` with a semaphore making sure the number of elements actually in the queue never rises beyond the capacity.

Such a queue would not guarantee that no allocations occur, but they should occur rarely during the lifetime of a queue; only happening a couple of times in the booting phase of a producer, since there should be a few extra blocks available in each producer when a block that has not been dequeued on becomes available for enqueue.

I did find an implementation with both capabilities in boost, albeit marked as "experimental" and since I suspect you are not terribly keen on having a less rigorous queue in terms of malloc calls in here I think I'll see if I can use that, even though there is a good chance it would beat the boost implementation :P

I'll see if the boost queue works and if it does I'll close the ticket…

Thank you so much for looking at the code and writing down your thoughts! I learned a bunch!

cameron314 commented:
Hey, thank *you* for daring to dive into somebody else's complex lock-free code! :-)

I agree that the current queue architecture is not conducive to these features and that trying to jump through hoops to add them is not worth it.

I'd like to write a concurrent ring buffer sometime Soon™ (i.e. probably in several months to years, heh). With a fixed size, blocking enqueue and dequeue can be easily overlaid, and close() is less necessary since the operations would be fully linearizable (i.e. close() would boil down to just an atomic flag set).

koraa commented:
OK. Feel free to tell me if you need any help with the ring buffer then :)
